src/ray/raylet/node_manager.cc

```diff
@@ -539,6 +545,66 @@ ray::Status NodeManager::RegisterGcs() {
         "NodeManager.deadline_timer.print_event_loop_stats");
   }
 
+  if (RayConfig::instance().use_ray_syncer()) {
+    // Register resource manager and scheduler
+    ray_syncer_.Register(
+        syncer::RayComponentId::RESOURCE_MANAGER,
+        this,
+        this,
+        false,
+        RayConfig::instance().raylet_report_resources_period_milliseconds());
+    ray_syncer_.Register(syncer::RayComponentId::SCHEDULER,
+                         this,
+                         nullptr,
+                         true,
+                         RayConfig::instance().raylet_report_loads_period_milliseconds());
+    // Register a commands channel.
+    // It's only used for GC right now.
+    ray_syncer_.Register(syncer::RayComponentId::COMMANDS, nullptr, this, false, 0);
+    periodical_runner_.RunFnPeriodically(
+        [this] {
+          // If plasma store is under high pressure, we should try to schedule a global
+          // gc.
+          bool plasma_high_pressure =
+              object_manager_.GetUsedMemoryPercentage() > high_plasma_storage_usage_;
+          if (plasma_high_pressure && global_gc_throttler_.AbleToRun()) {
+            TriggerGlobalGC();
+          }
+
+          // Set the global gc bit on the outgoing heartbeat message.
+          bool triggered_by_global_gc = false;
+          if (should_global_gc_) {
+            triggered_by_global_gc = true;
+            should_global_gc_ = false;
+            global_gc_throttler_.RunNow();
+          }
+
+          // Trigger local GC if needed. This throttles the frequency of local GC calls
+          // to at most once per heartbeat interval.
+          if ((should_local_gc_ || (absl::GetCurrentTimeNanos() - local_gc_run_time_ns_ >
+                                    local_gc_interval_ns_)) &&
+              local_gc_throttler_.AbleToRun()) {
+            DoLocalGC(triggered_by_global_gc);
+            should_local_gc_ = false;
+          }
+
+          if (triggered_by_global_gc) {
+            rpc::ResourcesData resources_data;
+            resources_data.set_should_global_gc(true);
+            syncer::RaySyncMessage msg;
+            msg.set_version(absl::GetCurrentTimeNanos());
+            msg.set_node_id(self_node_id_.Binary());
+            msg.set_component_id(syncer::RayComponentId::COMMANDS);
+            std::string serialized_msg;
+            RAY_CHECK(resources_data.SerializeToString(&serialized_msg));
```

Comment:
It looks like RaySyncMessage always need a bytes, so it means for every type of message we would do the ser/deser, will it become a potiential overload? (more conern about the GCS and the heavey messages)

Reply:
The current GCS do it as well. Whenever when you send a message it does a serialization. 
We trade copy for deserialization.